### PR TITLE
Non-nullable fields of default struct value should be treated as nullable

### LINKED ISSUE #30731: Non-nullable field of default struct value should be treated as nullable
Currently, the compiler warns when dereferencing `s.F1` but not when dereferencing `s.F2`:
```c#
#pragma warning disable 649

struct S
{
    internal object? F1;
    internal object F2;
}

class Program
{
    static void Main()
    {
        S s = default;
        s.F1.ToString(); // warning: possibly null
        s.F2.ToString(); // warning: possibly null
    }
}
```

### PR DESCRIPTION
**Pull request by cston**
Fixes #30731

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -17481,6 +17484,77 @@ static void F2(bool b, IOut<object, string> x2, IOut<object?, string?> y2)
             comp.VerifyTypes();
         }
 
+#if false
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Is this intentional? If it is please add a comment and link to follow-up issue.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -20775,6 +20849,65 @@ static void G3(B<object>.IOut<string> x3, B<object?>.IOut<string?> y3)
...
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Same here

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.VariableIdentifier.cs [hunk 3]
```diff
@@ -49,7 +48,7 @@ public override bool Equals(object obj)
                 return !left.Equals(right);
             }
 
-            internal string GetDebuggerDisplay()
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.VariableIdentifier.cs]: >GetDebuggerDisplay [](start = 28, length = 18)

I'm curious, why the change?
Just want to confirm it's not because `DebuggerDisplay` isn't working in dev16 (I've already filled a bug on that).

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 4]
```diff
@@ -38967,6 +38970,493 @@ static void F(object? x, object? y)
...
+        where U : class
+        where V : struct
+    {
+        default(S<T>).F/*T:T*/.ToString(); // 1
+        default(S<U>).F/*T:U?*/.ToString(); // 2
+        _ = default(S<V?>).F/*T:V?*/.Value; // 3
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            // PROTOTYPE: Missing // 2
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >PROTOTYPE [](start = 15, length = 9)

Need to log a bug if not resolved in this PR.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 5]
```diff
@@ -38967,6 +38970,493 @@ static void F(object? x, object? y)
...
+@"class Program
+{
+    static void F()
+    {
+        default((object?, string))/*T:(object?, string!)*/.Item2.ToString(); // 1
+        _ = default((int, int?))/*T:(int, int?)*/.Item2.Value; // 2
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            // PROTOTYPE: Missing // 1
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >PROTOTYPE [](start = 15, length = 9)

Same issue as earlier.

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.VariableIdentifier.cs (hunk 3 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.VariableIdentifier.cs]: Changed because `ToString()` is used in the EE for instances in a collection.

File: docs/features/nullable-reference-types.md [hunk 6]
```diff
@@ -94,14 +94,25 @@ Invocation of methods annotated with the following attributes will also affect f
...
-`default(T)` is `T?` if `T` is a reference type.
-_Is `default(T)` also `T?` if `T` is an unconstrained type parameter?_
-_Is `default(T?)` an error?_
+If `T` is a reference type, `default(T)` is `T?`.
 ```c#
 string? s = default(string); // assigns ?, no warning
 string t = default; // assigns ?, warning
-T t = default; // assigns ?, warning
 ```
+If `T` is a value type, `default(T)` is `T` and any non-nullable fields in `T` are maybe null.
```
Comment (jcouv) [inline docs/features/nullable-reference-types.md]: >any non-nullable fields [](start = 48, length = 23)

Perhaps "all non-struct fields"?
`int?`, `string?` or `T` field would also get a nullable null-state, I assume.

File: docs/features/nullable-reference-types.md [hunk 7]
```diff
@@ -94,14 +94,25 @@ Invocation of methods annotated with the following attributes will also affect f
...
+If `T` is a value type, `new T()` is equivalent to `default(T)`.
```
Comment (jcouv) [inline docs/features/nullable-reference-types.md]: >`new T()` is equivalent to `default(T)` [](start = 24, length = 39)

Does this depend on whether we're using a default constructor, or a user-defined one?

Comment (cston) [reply docs/features/nullable-reference-types.md]: This only applies to the default parameter-less constructor that cannot be overridden for structs.

---
In reply to: [246863817](https://github.com/dotnet/roslyn/pull/32304#discussion_r246863817) [](ancestors = 246863817)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 8]
```diff
@@ -797,7 +831,9 @@ private void InheritNullableStateOfMember(int targetContainerSlot, int valueCont
             if (fieldOrPropertyType.IsReferenceType || fieldOrPropertyType.IsNullableType())
             {
                 int targetMemberSlot = GetOrCreateSlot(member, targetContainerSlot);
-                NullableAnnotation value = fieldOrPropertyType.NullableAnnotation;
+                NullableAnnotation value = (isDefaultValue && fieldOrPropertyType.IsReferenceType) ?
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >fieldOrPropertyType.IsReferenceType [](start = 62, length = 35)

Should this be `... && (fieldOrPropertType.IsValueType || fieldOrProperty.IsNullableType())`?

Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: If the member type is a nullable value type, `fieldOrPropertyType.NullableAnnotation` should return `Nullable` already.

---
In reply to: [246866495](https://github.com/dotnet/roslyn/pull/32304#discussion_r246866495) [](ancestors = 246866495)

File: docs/features/nullable-reference-types.md (hunk 7 above)
Comment (jcouv) [reply docs/features/nullable-reference-types.md]: Ah, I didn't know they couldn't be overridden. Thanks

---
In reply to: [246865325](https://github.com/dotnet/roslyn/pull/32304#discussion_r246865325) [](ancestors = 246865325,246863817)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 9]
```diff
@@ -1290,7 +1329,7 @@ private void SetResult(BoundExpression node)
             _resultType = TypeSymbolWithAnnotations.Create(node.Type);
         }
 
-        private ObjectCreationPlaceholderLocal GetOrCreateObjectCreationPlaceholder(BoundExpression node)
+        private int GetOrCreateObjectCreationPlaceholder(BoundExpression node)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >GetOrCreateObjectCreationPlaceholder [](start = 20, length = 36)

Perhaps GetOrCreateObjectCreationPlaceholderSlot ?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 10]
```diff
@@ -928,13 +964,11 @@ private void EnterParameter(ParameterSymbol parameter, TypeSymbolWithAnnotations
             {
                 if (EmptyStructTypeCache.IsTrackableStructType(parameterType.TypeSymbol))
                 {
-                    InheritNullableStateOfTrackableStruct(parameterType.TypeSymbol, slot, valueSlot: -1, isByRefTarget: parameter.RefKind != RefKind.None, slotWatermark: GetSlotWatermark());
+                    InheritNullableStateOfTrackableStruct(parameterType.TypeSymbol, slot, valueSlot: -1, isDefaultValue: false, isByRefTarget: parameter.RefKind != RefKind.None, slotWatermark: GetSlotWatermark());
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >isDefaultValue: false, [](start = 105, length = 22)

What about `void M(string s = default)`?
I think we report a warning already, but in terms of null-state the optional value should probably still count.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 11]
```diff
@@ -1090,7 +1124,7 @@ private static bool IsTypeParameterDisallowingAnnotation(TypeSymbol typeOpt)
         public override BoundNode VisitLocal(BoundLocal node)
         {
             var local = node.LocalSymbol;
-            int slot = GetOrCreateSlot(local);
+            int slot = GetOrCreateSlot(local, createIfMissing: false);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >, createIfMissing: false [](start = 44, length = 24)

Does this matter? Consider removing for simplicity.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 8 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: A member with a unconstrained type parameter type is already considered maybe null. The only cases we need to handle specifically here are members where the type is declared as not nullable.

---
In reply to: [246871048](https://github.com/dotnet/roslyn/pull/32304#discussion_r246871048) [](ancestors = 246871048,246867516,246866495)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 11 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: This shouldn't matter (since a slot for the local should have been created already), but it seems reasonable to pass `false` explicitly so the intent is clear that a slot should not be created here.

---
In reply to: [246872573](https://github.com/dotnet/roslyn/pull/32304#discussion_r246872573) [](ancestors = 246872573)

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 8 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: There is a related issue when copying not-nullable state for unconstrained types though. Logged [https://github.com/dotnet/roslyn/issues/32338](https://github.com/dotnet/roslyn/issues/32338). Thanks.

---
In reply to: [246874741](https://github.com/dotnet/roslyn/pull/32304#discussion_r246874741) [](ancestors = 246874741,246871048,246867516,246866495)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 10 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Good catch, thanks.

---
In reply to: [246871918](https://github.com/dotnet/roslyn/pull/32304#discussion_r246871918) [](ancestors = 246871918)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 12]
```diff
@@ -928,13 +964,17 @@ private void EnterParameter(ParameterSymbol parameter, TypeSymbolWithAnnotations
             {
                 if (EmptyStructTypeCache.IsTrackableStructType(parameterType.TypeSymbol))
                 {
-                    InheritNullableStateOfTrackableStruct(parameterType.TypeSymbol, slot, valueSlot: -1, isByRefTarget: parameter.RefKind != RefKind.None, slotWatermark: GetSlotWatermark());
+                    InheritNullableStateOfTrackableStruct(
+                        parameterType.TypeSymbol,
+                        slot,
+                        valueSlot: -1,
+                        isDefaultValue: parameter.ExplicitDefaultConstantValue?.IsNull == true,
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >parameter.ExplicitDefaultConstantValue?.IsNull == true, [](start = 40, length = 55)

Does that work for `void M(T t = default)`?

File: src/Compilers/CSharp/Portable/FlowAnalysis/LocalDataFlowPass.cs [hunk 13]
```diff
@@ -91,11 +91,11 @@ protected int VariableSlot(Symbol symbol, int containingSlot = 0)
         /// <summary>
         /// Force a variable to have a slot.  Returns -1 if the variable has an empty struct type.
         /// </summary>
-        protected int GetOrCreateSlot(Symbol symbol, int containingSlot = 0)
+        protected int GetOrCreateSlot(Symbol symbol, int containingSlot = 0, bool createIfMissing = true)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/LocalDataFlowPass.cs]: >createIfMissing [](start = 82, length = 15)

It feels like the "createIfMissing" parameter is specific to tuples, but the name doesn't reflect the fact. This might be confusing to consumers. For example, one might assume that, unless we already allocated a slot, this function won't allocate it when false is passed for ```createIfMissing``` parameter. However, it doesn't look like this is the case.

File: src/Compilers/CSharp/Portable/FlowAnalysis/LocalDataFlowPass.cs [hunk 14]
```diff
@@ -162,16 +162,10 @@ private int DescendThroughTupleRestFields(ref Symbol symbol, int containingSlot,
                         return -1;
                     }
 
-                    if (forceContainingSlotsToExist)
+                    containingSlot = GetOrCreateSlot(restField, containingSlot, createIfMissing);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/LocalDataFlowPass.cs]: >containingSlot = GetOrCreateSlot(restField, containingSlot, createIfMissing); [](start = 20, length = 77)

Is propagating ```createIfMissing``` has any real effect here? It looks like ```restField``` is coming from an underlying named type. Therefore, it cannot be a TupleFieldSymbol. Therefore, the ```DescendThroughTupleRestFields``` call in the ```GetOrCreateSlot``` above will be a NoOp regardless of the value passed in ```createIfMissing```.
If these observations are correct, then it means that now ```createIfMissing``` has no effect on the "result" of this function, but it looks like the ```forceContainingSlotsToExist``` parameter used to have an effect.

File: src/Compilers/CSharp/Portable/FlowAnalysis/LocalDataFlowPass.cs [hunk 15]
```diff
@@ -82,7 +82,7 @@ protected override void Free()
         /// </summary>
         protected int VariableSlot(Symbol symbol, int containingSlot = 0)
         {
-            containingSlot = DescendThroughTupleRestFields(ref symbol, containingSlot, forceContainingSlotsToExist: false);
+            containingSlot = DescendThroughTupleRestFields(ref symbol, containingSlot, createIfMissing: false);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/LocalDataFlowPass.cs]: >createIfMissing: false [](start = 87, length = 22)

Does passing false really makes any difference given the changes in ```DescendThroughTupleRestFields``` function?

File: src/Compilers/CSharp/Portable/FlowAnalysis/LocalDataFlowPass.cs [hunk 16]
```diff
@@ -91,11 +91,11 @@ protected int VariableSlot(Symbol symbol, int containingSlot = 0)
         /// <summary>
         /// Force a variable to have a slot.  Returns -1 if the variable has an empty struct type.
         /// </summary>
-        protected int GetOrCreateSlot(Symbol symbol, int containingSlot = 0)
+        protected int GetOrCreateSlot(Symbol symbol, int containingSlot = 0, bool createIfMissing = true)
         {
             if (symbol.Kind == SymbolKind.RangeVariable) return -1;
 
-            containingSlot = DescendThroughTupleRestFields(ref symbol, containingSlot, forceContainingSlotsToExist: true);
+            containingSlot = DescendThroughTupleRestFields(ref symbol, containingSlot, createIfMissing);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/LocalDataFlowPass.cs]: >containingSlot = DescendThroughTupleRestFields(ref symbol, containingSlot, createIfMissing); [](start = 12, length = 92)

I understand that this is an existing logic. It looks like ```DescendThroughTupleRestFields``` can return ```-1```. Does it make sense to continue in this case? It looks like we will go ahead and allocate a new slot for the member even in this case. It also looks like that slot is going to be shared across distinct original containers, effectively treating members of different containers as the same storage location.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 17]
```diff
@@ -436,6 +436,21 @@ protected override int MakeSlot(BoundExpression node)
                         }
                     }
                     break;
+                case BoundKind.DefaultExpression:
+                    {
+                        var type = ((BoundDefaultExpression)node).Type;
+                        if (EmptyStructTypeCache.IsTrackableStructType(type))
+                        {
+                            int slot = GetOrCreateObjectCreationPlaceholderSlot(node);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >int slot = GetOrCreateObjectCreationPlaceholderSlot(node); [](start = 28, length = 58)

It feels like there is an inconsistency between when we create placeholder locals for the ```default``` expression and for an object creation. I think it would be better to use the same approach for all such cases. For example, here we would simply do what is done for the ```case BoundKind.ObjectCreationExpression:``` and the code that is currently here could be moved to ```VisitDefault``` etc.

Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Yes, that is necessary to handle split states as well.

---
In reply to: [246945160](https://github.com/dotnet/roslyn/pull/32304#discussion_r246945160) [](ancestors = 246945160)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 18]
```diff
@@ -445,15 +460,16 @@ protected override int MakeSlot(BoundExpression node)
...
                 case BoundKind.ConditionalReceiver:
-                    if (_lastConditionalAccessSlot != -1)
                     {
                         int slot = _lastConditionalAccessSlot;
                         _lastConditionalAccessSlot = -1;
                         return slot;
                     }
-                    break;
+                default:
+                    return base.MakeSlot(node);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >return base.MakeSlot(node); [](start = 20, length = 27)

This used to be called for all cases that don't ```return```, now we always return ```-1``` for scenarios like that. Is this the right thing to do?

Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Yes, I believe so. There is no overlap between the cases here and in the base.

---
In reply to: [246946078](https://github.com/dotnet/roslyn/pull/32304#discussion_r246946078) [](ancestors = 246946078)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 19]
```diff
@@ -602,22 +618,39 @@ bool reportNullLiteralAssignmentIfNecessary(BoundExpression expr)
                 }
                 return true;
             }
+        }
 
-            bool isDefaultOfUnconstrainedTypeParameter(BoundExpression expr)
+        private static bool IsDefaultOfUnconstrainedTypeParameter(BoundExpression expr)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >IsDefaultOfUnconstrainedTypeParameter [](start = 28, length = 37)

It looks like this function is called only from the original call site. Why make it a member vs. a local function?

Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: To allow grouping the two `IsDefault...` methods so the similarity is more obvious.

---
In reply to: [246948367](https://github.com/dotnet/roslyn/pull/32304#discussion_r246948367) [](ancestors = 246948367)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 12 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: We're only interested in cases where `parameter.Type` is a trackable `struct`.

---
In reply to: [246931559](https://github.com/dotnet/roslyn/pull/32304#discussion_r246931559) [](ancestors = 246931559)

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >isDefaultValue: parameter.ExplicitDefaultConstantValue?.IsNull == true, [](start = 24, length = 71)

Is this going to properly handle structures?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 20]
```diff
@@ -832,13 +868,13 @@ private void InheritNullableStateOfMember(int targetContainerSlot, int valueCont
                     int valueMemberSlot = -1;
                     if (valueContainerSlot > 0)
                     {
-                        int slot = GetOrCreateSlot(member, valueContainerSlot);
+                        int slot = GetOrCreateSlot(member, valueContainerSlot, createIfMissing: false);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >createIfMissing: false [](start = 79, length = 22)

I am not sure if passing this value is going to have any useful effect. If we are not interested in allocating new slots, perhaps we should use ```VariableSlot``` API instead.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 21]
```diff
@@ -1090,7 +1130,7 @@ private static bool IsTypeParameterDisallowingAnnotation(TypeSymbol typeOpt)
         public override BoundNode VisitLocal(BoundLocal node)
         {
             var local = node.LocalSymbol;
-            int slot = GetOrCreateSlot(local);
+            int slot = GetOrCreateSlot(local, createIfMissing: false);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >createIfMissing: false [](start = 46, length = 22)

I am not sure if passing this value is going to have any useful effect. If we are not interested in allocating new slots, perhaps we should use VariableSlot API instead.

**Comment by AlekseyTs** [comment]
>             LocalSymbol receiver = null;

It looks like receiver is never changed to anything else.

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:1225 in 50416ca. [](commit_id = 50416caf6379d7fa64864c18d37edcafa0a787c3, deletion_comment = False)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 22]
```diff
@@ -1190,12 +1230,17 @@ private void VisitObjectOrDynamicObjectCreation(BoundExpression node, BoundExpre
...
+                    slot = GetOrCreateObjectCreationPlaceholderSlot(node);
                     if (slot > 0 && isTrackableStructType)
                     {
                         this.State[slot] = NullableAnnotation.NotNullable;
-                        InheritNullableStateOfTrackableStruct(type, slot, valueSlot: -1, isByRefTarget: false, slotWatermark: GetSlotWatermark());
+                        InheritNullableStateOfTrackableStruct(
+                            type,
+                            slot,
+                            valueSlot: -1,
+                            isDefaultValue: (node as BoundObjectCreationExpression)?.Constructor.IsImplicitConstructor == true,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >(node as BoundObjectCreationExpression)?.Constructor.IsImplicitConstructor == true [](start = 44, length = 82)

This doesn't look like a test for a Default Value Type Constructor. Am I misinterpreting the intent?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 23]
```diff
@@ -3940,7 +3984,7 @@ private void VisitThisOrBaseReference(BoundExpression node)
         public override BoundNode VisitParameter(BoundParameter node)
         {
             var parameter = node.ParameterSymbol;
-            int slot = GetOrCreateSlot(parameter);
+            int slot = GetOrCreateSlot(parameter, createIfMissing: false);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >createIfMissing: false [](start = 50, length = 22)

I am not sure if passing this value is going to have any useful effect. If we are not interested in allocating new slots, perhaps we should use VariableSlot API instead.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 24]
```diff
@@ -4301,7 +4345,7 @@ private void VisitMemberAccess(BoundExpression node, BoundExpression receiverOpt
                     if (!resultType.IsValueType || resultType.IsNullableType())
                     {
                         int containingSlot = getReceiverSlot();
-                        int slot = (containingSlot < 0) ? -1 : GetOrCreateSlot(member, containingSlot);
+                        int slot = (containingSlot < 0) ? -1 : GetOrCreateSlot(member, containingSlot, createIfMissing: false);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >createIfMissing: false [](start = 103, length = 22)

I am not sure if passing this value is going to have any useful effect. If we are not interested in allocating new slots, perhaps we should use VariableSlot API instead. But do we really want to skip allocating the slot and fail to adjust ```resultType```?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 12 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: `parameter.ExplicitDefaultConstantValue.IsNull` is `true` for `default(S)` and `new S()`.

---
In reply to: [246951009](https://github.com/dotnet/roslyn/pull/32304#discussion_r246951009) [](ancestors = 246951009)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 25]
```diff
@@ -771,7 +804,7 @@ private void ReportDiagnostic(ErrorCode errorCode, SyntaxNode syntaxNode, params
             }
         }
 
-        private void InheritNullableStateOfTrackableStruct(TypeSymbol targetType, int targetSlot, int valueSlot, bool isByRefTarget, int slotWatermark)
+        private void InheritNullableStateOfTrackableStruct(TypeSymbol targetType, int targetSlot, int valueSlot, bool isDefaultValue, bool isByRefTarget, int slotWatermark)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >bool isDefaultValue [](start = 113, length = 19)

If we are propagating this flag separately anyway, why do we need slots for the "fake" state?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 26]
```diff
@@ -38967,6 +38970,782 @@ static void F(object? x, object? y)
...
+{
+    internal T F;
+}
+class Program
+{
+    static void F<T, U, V>()
+        where U : class
+        where V : struct
+    {
+        new S<T>().F/*T:T*/.ToString(); // 1
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >new S<T>() [](start = 8, length = 10)

Do we have a test covering non-default value type constructors and explicitly defined parameter-less constructors (legal in metadata I think)?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 22 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Changed.

---
In reply to: [246953718](https://github.com/dotnet/roslyn/pull/32304#discussion_r246953718) [](ancestors = 246953718)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 27]
```diff
@@ -38967,6 +38970,782 @@ static void F(object? x, object? y)
...
+        new S<U>().F/*T:U?*/.ToString(); // 2
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >new S<U>() [](start = 8, length = 10)

Do we have a test covering combination of a default value type constructors with object member initializers?

**Comment by AlekseyTs** [comment]
Done review pass (iteration 11)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 12 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Added a missing default value test, and checked the implementations of `ExplicitDefaultConstantValue` look reasonable.

---
In reply to: [246959155](https://github.com/dotnet/roslyn/pull/32304#discussion_r246959155) [](ancestors = 246959155,246955015,246951009)
